Replace winapi with windows-sys

vthib opened the pull request:
[windows-sys](https://github.com/microsoft/windows-rs) is a crate with bindings to the windows API that is maintained by Microsoft. A lot of crates are migrating from `winapi` to it, and it would be nice for `libloading` to migrate as well.

The changes are quite straightforward:

- the DWORD, WCHAR, etc types are not exposed by windows-sys, the standard types are used instead: u32, u16, etc.
- HMODULE does not exist, HINSTANCE is used instead, and is a `isize` instead of a pointer.
- FARPROC is a `Option<unsafe extern "system" fn() -> isize>`, which is a bit more annoying to use in debug prints: I used the `{:#x}` formatter instead of the `{:p}` one: before: ``, after: `Symbol@0x7ffb29135bf0`.


Closes #118

nagisa commented on src/os/windows/mod.rs:
```diff
@@ -186,16 +183,15 @@ impl Library {
     /// termination routines contained within the library is safe as well. These routines may be
     /// executed when the library is unloaded.
-    pub unsafe fn load_with_flags<P: AsRef<OsStr>>(filename: P, flags: DWORD) -> Result<Library, crate::Error> {
+    pub unsafe fn load_with_flags<P: AsRef<OsStr>>(filename: P, flags: u32) -> Result<Library, crate::Error> {
```
I’m not sure I’m particularly comfortable with this change. Ultimately `LoadLibraryExW` is accepting a DWORD and I have no way of ensuring that Microsoft won’t wake up one of these days and produce a Windows port with `DWORD != u32`. They have a long history of making some real difficult to understand choices after all.

`windows-sys` appears to have a [type alias](https://docs.rs/windows-sys/latest/windows_sys/Win32/System/LibraryLoader/type.LOAD_LIBRARY_FLAGS.html) for this argument. Is there a reason to not use that?

vthib replied:
fwiw, this is discussed a bit in the windows-rs repo: https://github.com/microsoft/windows-rs/issues/881 where they recommend to just use the underlying value.

Using the type alias is however a very good idea, done

nagisa commented on src/os/windows/mod.rs:
```diff
@@ -2,38 +2,35 @@
-    pub(super) use self::winapi::um::{errhandlingapi, libloaderapi};
+    extern crate windows_sys;
+    pub(super) use self::windows_sys::Win32::Foundation::{GetLastError, FARPROC, HINSTANCE};
+    pub(super) use self::windows_sys::Win32::System::Diagnostics::Debug as DebugApi;
+    pub(super) use self::windows_sys::Win32::System::LibraryLoader;
```
```suggestion
    pub(super) use self::windows_sys::Win32::System::Diagnostics::Debug as debug_api;
    pub(super) use self::windows_sys::Win32::System::LibraryLoader as library_loader;
```

vthib replied:
Done

nagisa commented on src/os/windows/mod.rs:
```diff
@@ -365,25 +361,30 @@ impl<T> ::std::ops::Deref for Symbol<T> {
 impl ErrorModeGuard {
     #[allow(clippy::if_same_then_else)]
     fn new() -> Option<ErrorModeGuard> {
+        use self::DebugApi::SEM_FAILCRITICALERRORS as SEM_FAILCE;
```
I’d say import this where the `SEM_FAILCE` was defined originally, and maybe drop the rename?

vthib replied:
Done

nagisa reviewed (commented):
Thanks for doing this, some comments inline. Some of them apply more broadly than just to the line indicated, for example the one about using the type alias in `load_with_flags` applies just as well to the flag (re-)definitions as well.

nagisa commented on src/os/windows/mod.rs:
```diff
@@ -305,17 +301,17 @@ impl fmt::Debug for Library {
                 buf[..].as_mut_ptr().cast(), 1024) as usize;
             if len == 0 {
-                f.write_str(&format!("Library@{:p}", self.0))
+                f.write_str(&format!("Library@{:#x}", self.0))
```
Ah, never mind, you explained this in the PR description.

vthib replied:
the # in `:#x` makes it output with a `0x` prefix. However I'm not sure if there aren't differences in for example the number of digits being printed, so this is very likely a non functional change yes

vthib commented:
@nagisa can you take another look?

nagisa commented:
Sorry for the delay, it looks like the rustdoc tests are failing. See CI results.

vthib commented:
I have fixed the rustdoc issues. There is however a clippy one, that stems from a windows metadata bug: https://github.com/microsoft/win32metadata/issues/1505

This is quite unfortunate, we'll have to wait for a new windows_sys release before we can merge this MR.

nagisa commented:
Looks like there is still a failure (and thus the release of windows-sys hasn't happened yet). Please ping me when the release occurs, thanks!

vthib commented:
@nagisa windows-sys released the fix on 0.48, should be working fine now!

nagisa commented:
Looks like there are some tests that are failing, but we will still need to raise the [minimum supported Rust version checked by the CI](https://github.com/nagisa/rust_libloading/blob/master/.github/workflows/libloading.yml#L18).

vthib commented:
Indeed, sorry about that. I have fixed the issues, I did not handle the Deref impl on the Symbol object properly.

I also added a MSRV bump to 1.48, this should fix the MSRV pipeline as well. I think everything should be fine now

nagisa commented:
Thanks!

MarijnS95 commented:
Fwiw the status-quo is to use `windows-bindgen` nowadays when accessing only a minimal `windows-sys` subset, rather than pulling in that massive crate.

https://github.com/microsoft/windows-rs/issues/1720#issuecomment-1497497708

MarijnS95 commented:
> Yes, discussion about this was ongoing while this PR was already in progress.

Curious where that's at, it's not in #118 nor the comments here.

> `windows-sys` is not that massive

Clocking in at 2.63MB it is by far the largest crate in most of my dependency graphs, while only a microscopic portion of its API surface is used.

vthib commented:
> Curious where that's at, it's not in #118 nor the comments here.

I meant this discussion: https://github.com/microsoft/windows-rs/issues/1720 which started again about two weeks ago.
We have not discussed the option of using `windows-bindgen` for `libloading`.

riverar commented:
> Fwiw the status-quo is to use `windows-bindgen` nowadays when accessing only a minimal `windows-sys` subset, rather than pulling in that massive crate.

It's a two week old feature with ~1 user and only designed for the most demanding of scenarios. I don't think you should be referring to this as the _status quo_ (or at all).

MarijnS95 commented:
It looked as if some high-value crates were getting this conversion to `windows-bindgen` as their API usage has been extremely light to warrant pulling in the massive-in-comparison `windows-sys` crate, but those PRs have been closed now.

ErichDonGubler commented:
Another interesting thing to note is that `parking_lot` has taken to using `windows-targets` directly, because they considered even the size of `windows-sys` intolerably high for their use case as a cornerstone crate of Rust's ecosystem.

I don't know what tolerances are here for code size of dependencies in this project vs. manual bindings work. Just examining this PR, it seems that there's a significantly higher number of APIs being used than [`parking-lot`'s 5 functions plus a dozen or so data types and constants](https://github.com/Amanieu/parking_lot/blob/e9f9f40b860d7de8238a8c07aee3ae05992ab681/core/src/thread_parker/windows/bindings.rs).

To be clear: I'm not proposing any action items for the project. Just wanted to raise something that might be relevant.
